src/nvim/eval/decode.c

```diff
@@ -0,0 +1,1094 @@
+            fst_in_pair = (int) ch;
+          } else if (SURROGATE_LO_START <= ch && ch <= SURROGATE_LO_END
+                     && fst_in_pair != 0) {
+            const int full_char = (
+                (int) (ch - SURROGATE_LO_START)
+                + ((fst_in_pair - SURROGATE_HI_START) << 10)
+                + SURROGATE_FIRST_CHAR);
+            str_end += utf_char2bytes(full_char, (char_u *) str_end);
+            fst_in_pair = 0;
+          } else {
+            PUT_FST_IN_PAIR(fst_in_pair, str_end);
+            str_end += utf_char2bytes((int) ch, (char_u *) str_end);
+          }
+          break;
+        }
+        case '\\':
+        case '/':
+        case '"':
+        case 't':
+        case 'b':
+        case 'n':
+        case 'r':
+        case 'f': {
+          static const char escapes[] = {
+            ['\\'] = '\\',
+            ['/'] = '/',
+            ['"'] = '"',
+            ['t'] = TAB,
+            ['b'] = BS,
+            ['n'] = NL,
+            ['r'] = CAR,
+            ['f'] = FF,
+          };
+          *str_end++ = escapes[(int) *t];
+          break;
+        }
+        default: {
+          assert(false);
+        }
+      }
+    } else {
+      *str_end++ = *t;
+    }
+  }
+  PUT_FST_IN_PAIR(fst_in_pair, str_end);
+#undef PUT_FST_IN_PAIR
+  if (conv->vc_type != CONV_NONE) {
+    size_t str_len = (size_t) (str_end - str);
+    char *const new_str = (char *) string_convert(conv, (char_u *) str,
+                                                  &str_len);
+    if (new_str == NULL) {
+      emsgf(_("E474: Failed to convert string \"%.*s\" from UTF-8"),
+            (int) str_len, str);
+      xfree(str);
+      goto parse_json_string_fail;
+    }
+    xfree(str);
+    str = new_str;
+    str_end = new_str + str_len;
+  }
+  if (hasnul) {
+    typval_T obj;
+    list_T *const list = list_alloc();
+    list->lv_refcount++;
+    create_special_dict(&obj, kMPString, ((typval_T) {
+      .v_type = VAR_LIST,
+      .v_lock = VAR_UNLOCKED,
+      .vval = { .v_list = list },
+    }));
+    if (encode_list_write((void *) list, str, (size_t) (str_end - str))
+        == -1) {
+      clear_tv(&obj);
+      goto parse_json_string_fail;
+    }
+    xfree(str);
+    POP(obj, true);
+  } else {
+    *str_end = NUL;
+    POP(((typval_T) {
+      .v_type = VAR_STRING,
+      .vval = { .v_string = (char_u *) str },
+    }), false);
+  }
+  goto parse_json_string_ret;
+parse_json_string_fail:
+  ret = FAIL;
+parse_json_string_ret:
+  *pp = p;
+  return ret;
+}
+
+#undef POP
+
+/// Parse JSON number: both floating-point and integer
+///
+/// Number format: `-?\d+(?:.\d+)?(?:[eE][+-]?\d+)?`.
+///
+/// @param[in]  buf  Buffer being converted.
+/// @param[in]  buf_len  Length of the buffer.
+/// @param[in,out]  pp  Pointer to the start of the number. Must point to
+///                     a digit or a minus sign. Is advanced to the last
+///                     character of the number.
```

Comment:
09.03.2016, 01:34, "oni-link" <notifications@github.com>:In src/nvim/eval/decode.c:> +parse_json_string_ret:
> +  *pp = p;
> +  return ret;
> +}
> +
> +#undef POP
> +
> +/// Parse JSON number: both floating-point and integer
> +///
> +/// Number format: `-?\d+(?:.\d+)?(?:[eE][+-]?\d+)?`.
> +///
> +/// @param[in]  buf  Buffer being converted.
> +/// @param[in]  buf_len  Length of the buffer.
> +/// @param[in,out]  pp  Pointer to the start of the number. Must point to
> +///                     a digit or a minus sign. Is advanced to the last
> +///                     character of the number.
[out]: In case of an error this does not point to a number (1.) ? —Reply to this email directly or view it on GitHub. 